src/hash.rs

```diff
@@ -45,100 +155,146 @@ impl std::str::FromStr for Hash {
             .next()
             .ok_or_else(|| Error::ParseIntegrityError(s.into()))?
             .parse()?;
-        let digest = String::from(
-            parsed
-                .next()
-                .ok_or_else(|| Error::ParseIntegrityError(s.into()))?,
-        );
-        Ok(Hash { algorithm, digest })
+        let digest_str = parsed
+            .next()
+            .ok_or_else(|| Error::ParseIntegrityError(s.into()))?
+            .trim_end();
+        let digest = STANDARD_BASE64
+            .decode(digest_str)
+            .map_err(|e| Error::ParseIntegrityError(e.to_string()))?;
+        Ok(match algorithm {
+            Algorithm::Sha512 => {
+                let mut hash = [0; 64];
+                if digest.len() != hash.len() {
+                    return Err(Error::ParseIntegrityError(format!(
+                        "Parsed {} bytes of hash - a {} hash should be {} bytes long",
+                        digest.len(),
+                        algorithm,
+                        hash.len()
+                    )));
+                }
+                hash.copy_from_slice(&digest);
+                Hash::Sha512(hash)
```

Comment:
It annoys me that this is creating a `Vec<u8>`, only to then copy its contents into an array, rather than just output it into the array. But calling `STANDARD_BASE64.decode_slice(digest_str, &mut hash)` returns `ParseIntegrityError("Output slice too small")` on valid input, because an internal size estimate is off. There is `decode_slice_unchecked`, but that on the other hand panics on too long input ("range end index 24 out of range for slice of length 20")

Reply:
Ah, this is fixed with a newer version. https://github.com/marshallpierce/rust-base64/blob/master/RELEASE-NOTES.md#0220